### ISSUE TITLE
Did you try to use the pretrained weight?

### ISSUE THREAD

**Issue by zzzucf** [PatrickZH/Improved-Deep-Leakage-from-Gradients#4]
Hi, I wonder did you try to use the pretrained weight on cifar rather than randomly initialized weights?

**Comment by PatrickZH** [comment]
Yes. I remember it works for pretrained weights. Learning rate may need to be adjusted.

**Comment by zzzucf** [comment]
Can you offer a case? And can you provide the pytorch version you used?

**Comment by PatrickZH** [comment]
It's easy to implement. Just pre-train the network before learning images from gradients. I have tried that. For a pre-trained LeNet that has 95% acc on MNIST and 10% acc on CIFAR100, it's still easy to recover the CIFAR100 images. But, recovering MNIST images often fails. The difficulty is caused by the samll gradient magnitude produced by well-trained models. This is not surperising. So, better matching loss, optimizer and learning rate schedule are required. Finally, I attach the code for training networks for you. 


            dataloader = torch.utils.data.DataLoader(dst, batch_size=256, shuffle=True, drop_last=True, num_workers=2)
            criterion = nn.CrossEntropyLoss().to(device)
            optimizer = torch.optim.SGD(net.parameters(), lr=0.1, weight_decay=0.0005)
            print('\n------ pre-train network ------')
            for epoch in range(20):
                for im, lab in dataloader:
                    im, lab = im.to(device), lab.to(device)
                    output = net(im)
                    loss = criterion(output, lab)
                    acc = np.mean(np.argmax(output.detach().cpu().numpy(), axis=-1) == lab.detach().cpu().numpy())
                    optimizer.zero_grad()
                    loss.backward()
                    optimizer.step()
                print('epoch = %04d train loss = %.6f acc = %.4f '%(epoch, loss.item(), acc))
            print('------------------------- ')

**Comment by zzzucf** [comment]
This is the problem I found too. 

As you tested, you used 95% pre-trained weight on MNIST and find out that reconstruction from MNIST is difficult. While for Cifar 100, 10% is almost like not being trained for a few, I think it will act similar as random initialized weights. In comparison, I used a 66% Cifar 10 and it's really difficult to reconstruct images.

The real concern is that, CNN will turn into a **non-injective** function during training. In simple words, different input will lead to the same output. And in deep leakage, this phenomenon will be bad as several input will lead into the same gradient (After some layers,  the activation can be the same and the gradient will be the same). As a result, leakage will rarely happen.
